snapmodel_policy.go

```diff
@@ -0,0 +1,100 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+
+	"github.com/canonical/go-tpm2"
+	"github.com/snapcore/snapd/asserts"
+
+	"golang.org/x/xerrors"
+)
+
+// SnapModelProfileParams provides the parameters to AddSnapModelProfile.
+type SnapModelProfileParams struct {
+	// PCRAlgorithm is the algorithm for which to compute PCR digests for. TPMs compliant with the "TCG PC Client Platform TPM Profile
+	// (PTP) Specification" Level 00, Revision 01.03 v22, May 22 2017 are required to support tpm2.HashAlgorithmSHA1 and
+	// tpm2.HashAlgorithmSHA256. Support for other digest algorithms is optional.
+	PCRAlgorithm tpm2.HashAlgorithmId
+
+	// PCRIndex is the PCR that snap-bootstrap measures the model to.
+	PCRIndex int
+
+	// Models is the set of models to add to the PCR profile.
+	Models []*asserts.Model
+}
+
+// AddSnapModelProfile adds the snap model profile to the PCR protection profile, as measured by snap-bootstrap, in order to generate
+// a PCR policy that is bound to a specific set of device models. It is the responsibility of snap-bootstrap to verify the integrity
+// of the model that it has measured.
+//
+// The measurement of a model is a digest computed as follows (where H is the supplied params.PCRAlgorithm):
+//  digest1 = H(tpm2.HashAlgorithmSHA384 || sign-key-sha3-384 || brand-id)
+//  digest2 = H(digest1 || model)
```

Comment:
(series, brand-id, model) is supposed to be unique but this is really not controlled by anyone by the brands themselves, also an assertion can be revised (different assertion revision). As previously discussed because it controls some fundamental bits of snapd policy vs how it handles/what it allows wrt recovery seeds and encryption, I think we want also to measure  []byte(model.Grade())

Reply:
I've added the grade now, but converted it to a single byte enum to avoid having another variable length field to digest.